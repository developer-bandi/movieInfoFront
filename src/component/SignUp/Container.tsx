import useSignUp from './Hook';
import SignUpPresentational from './Presentational';

const SignUpContainer = () => {
  const { submitForm, setId, setPassword, setPasswordCheck, setNickname } =
    useSignUp();
  return (
    <SignUpPresentational
      setId={setId}
      setPassword={setPassword}
      setNickname={setNickname}
      setPasswordCheck={setPasswordCheck}
      submitForm={submitForm}
    />
  );
};

export default SignUpContainer;
