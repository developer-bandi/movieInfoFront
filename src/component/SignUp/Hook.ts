import axios from 'axios';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
var CryptoJS = require('crypto-js');

const useSignUp = () => {
  const [userid, setId] = useState('');
  const [password, setPassword] = useState('');
  const [passwordCheck, setPasswordCheck] = useState('');
  const [nick, setNickname] = useState('');
  //회원가입 폼 상태값은 컴포넌트 단에서 관리

  const navigate = useNavigate();
  const submitForm = async () => {
    try {
      const agreeSubmit = window.confirm(
        '가입시 닉네임은 변경할수 없습니다 가입하시겠습니까?'
      );
      if (agreeSubmit) {
        if (userid === '' || userid.length > 10) {
          alert('아이디를 정확히 입력하세요');
        } else if (password === '' || password.length > 10) {
          alert('비밀번호를 정확히 입력하세요');
        } else if (password !== passwordCheck) {
          alert('확인 비밀번호가 일치하지 않습니다 다시 입력해주세요');
        } else if (nick === '' || nick.length > 5) {
          alert('닉네임을 정확하게 입력하세요');
        } else {
          const userData = {
            userid,
            password: CryptoJS.AES.encrypt(
              JSON.stringify(password),
              process.env.REACT_APP_SECRET
            ).toString(),
            nick,
          };
          await axios.post(
            `${process.env.REACT_APP_SERVER}/auth/join`,
            userData
          );
          alert('회원가입에 성공하였습니다.');
          navigate('/signin');
        }
        //아이디 비밀번호 닉네임 조건으로 길이수를 설정하고 통과시에는 회원가입 완료처리
      }
    } catch (err) {
      alert('에러가 발생하였습니다');
    }
  };

  return {
    submitForm,
    setId,
    setPassword,
    setPasswordCheck,
    setNickname,
  };
};

export default useSignUp;
