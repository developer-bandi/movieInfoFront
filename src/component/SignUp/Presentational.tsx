import React from 'react';
import styles from './Style';

interface SignUpProps {
  setId: React.Dispatch<React.SetStateAction<string>>;
  setPassword: React.Dispatch<React.SetStateAction<string>>;
  setPasswordCheck: React.Dispatch<React.SetStateAction<string>>;
  setNickname: React.Dispatch<React.SetStateAction<string>>;
  submitForm: () => void;
}
const SignUpPresentational = ({
  setId,
  setPassword,
  setPasswordCheck,
  setNickname,
  submitForm,
}: SignUpProps) => {
  return (
    <styles.MainBlock>
      <styles.Title>회원가입</styles.Title>
      <section>
        <styles.InfoTitle>아이디</styles.InfoTitle>
        <styles.InfoInput
          placeholder="아이디를 10자리 미만으로 기입하세요"
          onChange={(e) => {
            setId(e.target.value);
          }}
          data-testid={'id'}
        ></styles.InfoInput>
      </section>
      <section>
        <styles.InfoTitle>비밀번호</styles.InfoTitle>
        <styles.InfoInput
          placeholder="비밀번호를 10자리 미만으로 기입하세요"
          onChange={(e) => {
            setPassword(e.target.value);
          }}
          type="password"
          data-testid={'password'}
        />
      </section>
      <section>
        <styles.InfoTitle>비밀번호 확인</styles.InfoTitle>
        <styles.InfoInput
          placeholder="위와 동일한 비밀번호를 입력하세요"
          onChange={(e) => {
            setPasswordCheck(e.target.value);
          }}
          type="password"
          data-testid={'passwordCheck'}
        />
      </section>
      <section>
        <styles.InfoTitle>닉네임</styles.InfoTitle>
        <styles.InfoInput
          placeholder="닉네임을 5자리 미만으로 기입하세요"
          onChange={(e) => {
            setNickname(e.target.value);
          }}
          data-testid={'nickname'}
        />
      </section>
      <styles.SubmitButton onClick={submitForm} data-testid={'submitButton'}>
        가입하기
      </styles.SubmitButton>
    </styles.MainBlock>
  );
};

export default SignUpPresentational;
