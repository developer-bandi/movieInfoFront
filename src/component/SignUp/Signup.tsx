import React from 'react';
import styles from './Style';

interface SignUpProps {
  setId: React.Dispatch<React.SetStateAction<string>>;
  setPassword: React.Dispatch<React.SetStateAction<string>>;
  setPasswordCheck: React.Dispatch<React.SetStateAction<string>>;
  setNickname: React.Dispatch<React.SetStateAction<string>>;
  submitForm: () => void;
}
const SignUp = ({
  setId,
  setPassword,
  setPasswordCheck,
  setNickname,
  submitForm,
}: SignUpProps) => {
  return (
    <styles.MainBlock>
      <styles.Title>회원가입</styles.Title>
      <section>
        <styles.InfoTitle>아이디</styles.InfoTitle>
        <styles.InfoInput
          placeholder="아이디를 10자리 미만으로 기입하세요"
          onChange={(e) => {
            setId(e.target.value);
          }}
        ></styles.InfoInput>
      </section>
      <section>
        <styles.InfoTitle>비밀번호</styles.InfoTitle>
        <styles.InfoInput
          placeholder="비밀번호를 10자리 미만으로 기입하세요"
          onChange={(e) => {
            setPassword(e.target.value);
          }}
          type="password"
        />
      </section>
      <section>
        <styles.InfoTitle>비밀번호 확인</styles.InfoTitle>
        <styles.InfoInput
          placeholder="위와 동일한 비밀번호를 입력하세요"
          onChange={(e) => {
            setPasswordCheck(e.target.value);
          }}
          type="password"
        />
      </section>
      <section>
        <styles.InfoTitle>닉네임</styles.InfoTitle>
        <styles.InfoInput
          placeholder="닉네임을 5자리 미만으로 기입하세요"
          onChange={(e) => {
            setNickname(e.target.value);
          }}
        />
      </section>
      <styles.SubmitButton onClick={submitForm}>가입하기</styles.SubmitButton>
    </styles.MainBlock>
  );
};

export default SignUp;
