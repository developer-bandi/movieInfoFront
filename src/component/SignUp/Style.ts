import styled from 'styled-components';

const MainBlock = styled.main`
  width: 21rem;
  height: 20.5rem;
  margin: auto;
  margin-top: 2rem;
`;

const Title = styled.h1`
  text-align: center;
  width: 100%;
  font-size: 35px;
  font-weight: bold;
  margin-top: 1rem;
  margin-bottom: 1rem;
`;

const InfoTitle = styled.h5`
  margin: 0;
  margin-bottom: 0.3rem;
`;

const InfoInput = styled.input`
  width: 19rem;
  height: 2.5rem;
  margin-bottom: 1rem;
  border: none;
  border: 1px solid black;
  border-radius: 5px;
  padding-left: 1rem;
`;

const SubmitButton = styled.div`
  width: 20.15rem;
  border-radius: 10px;
  height: 2.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  background: #39373a;
  color: white;
`;

const styles = { MainBlock, Title, InfoTitle, InfoInput, SubmitButton };

export default styles;
