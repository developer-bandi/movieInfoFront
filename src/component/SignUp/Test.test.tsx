import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from 'react-router-dom';
import useSignUp from './Hook';
import SignUp from './Presentational';

describe('SignIn Presentational', () => {
  it('전체 화면 체크', () => {
    const setIdMock = jest.fn();
    const setPasswordMock = jest.fn();
    const setPasswordCheckMock = jest.fn();
    const setNicknameMock = jest.fn();
    const submitFormMock = jest.fn();
    const utils = render(
      <SignUp
        setId={setIdMock}
        setPassword={setPasswordMock}
        setPasswordCheck={setPasswordCheckMock}
        setNickname={setNicknameMock}
        submitForm={submitFormMock}
      />,
      { wrapper: MemoryRouter }
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.change(screen.getByTestId('id'), {
      target: { value: 'new value' },
    });
    fireEvent.change(screen.getByTestId('password'), {
      target: { value: 'new value' },
    });
    fireEvent.change(screen.getByTestId('passwordCheck'), {
      target: { value: 'new value' },
    });
    fireEvent.change(screen.getByTestId('nickname'), {
      target: { value: 'new value' },
    });
    fireEvent.click(screen.getByTestId('submitButton'));

    expect(utils.container).toMatchSnapshot();
    expect(setIdMock).toBeCalledTimes(1);
    expect(setPasswordMock).toBeCalledTimes(1);
    expect(setPasswordCheckMock).toBeCalledTimes(1);
    expect(setNicknameMock).toBeCalledTimes(1);
    expect(submitFormMock).toBeCalledTimes(1);
  });
});

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('SignIn Hook', () => {
  afterEach(cleanup);
  window.alert = jest.fn();
  describe('checkLogin함수 테스트', () => {
    describe('아이디 입력 에러 발생', () => {
      it('아이디가 빈칸인 경우', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });

        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('아이디를 정확히 입력하세요');
      });

      it('아이디의 길이가 10 이상인 경우', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test1234567890');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('아이디를 정확히 입력하세요');
      });
    });

    describe('비밀번호 입력 에러 발생', () => {
      it('비밀번호가 빈칸인 경우', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('비밀번호를 정확히 입력하세요');
      });

      it('비밀번호 길이가 10 이상', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
          result.current.setPassword('test1234567890');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('비밀번호를 정확히 입력하세요');
      });
    });

    it('비밀번호 확인 입력 에러 발생', async () => {
      window.confirm = jest.fn(() => true);
      const { result } = renderHook(() => useSignUp(), {
        wrapper: MemoryRouter,
      });
      act(() => {
        result.current.setId('test');
        result.current.setPassword('test');
        result.current.setPasswordCheck('notEqule');
      });
      await act(async () => {
        await result.current.submitForm();
      });
      expect(window.alert).toBeCalledWith(
        '확인 비밀번호가 일치하지 않습니다 다시 입력해주세요'
      );
    });
    describe('닉네임 입력 에러 발생', () => {
      it('닉네임이 빈칸인 경우', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
          result.current.setPassword('test');
          result.current.setPasswordCheck('test');
        });

        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('닉네임을 정확하게 입력하세요');
      });
      it('닉네임 길이가 5이상인 경우', async () => {
        window.confirm = jest.fn(() => true);
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
          result.current.setPassword('test');
          result.current.setPasswordCheck('test');
          result.current.setNickname('test1234');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('닉네임을 정확하게 입력하세요');
      });
    });

    describe('회원 가입 조건 부합', () => {
      it('회원가입 성공', async () => {
        window.confirm = jest.fn(() => true);
        mockedAxios.post.mockImplementation(() =>
          Promise.resolve({
            status: 200,
            data: '회원가입 성공',
          })
        );
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
          result.current.setPassword('test');
          result.current.setPasswordCheck('test');
          result.current.setNickname('test');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('회원가입에 성공하였습니다.');
      });
      it('회원가입 에러 발생으로 실패', async () => {
        window.confirm = jest.fn(() => true);
        mockedAxios.post.mockImplementation(() =>
          Promise.reject({
            status: 500,
            data: '회원가입 실패',
          })
        );
        const { result } = renderHook(() => useSignUp(), {
          wrapper: MemoryRouter,
        });
        act(() => {
          result.current.setId('test');
          result.current.setPassword('test');
          result.current.setPasswordCheck('test');
          result.current.setNickname('test');
        });
        await act(async () => {
          await result.current.submitForm();
        });
        expect(window.alert).toBeCalledWith('에러가 발생하였습니다');
      });
    });
  });
});
