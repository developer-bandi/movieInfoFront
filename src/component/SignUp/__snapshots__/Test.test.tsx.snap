// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Presentational 전체 화면 체크 1`] = `
<div>
  <main
    class="sc-bdvvtL eNTxIt"
  >
    <h1
      class="sc-gsDKAQ hpyiGY"
    >
      회원가입
    </h1>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        아이디
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="id"
        placeholder="아이디를 10자리 미만으로 기입하세요"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        비밀번호
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="password"
        placeholder="비밀번호를 10자리 미만으로 기입하세요"
        type="password"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        비밀번호 확인
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="passwordCheck"
        placeholder="위와 동일한 비밀번호를 입력하세요"
        type="password"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        닉네임
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="nickname"
        placeholder="닉네임을 5자리 미만으로 기입하세요"
      />
    </section>
    <div
      class="sc-eCImPb kWtwVz"
      data-testid="submitButton"
    >
      가입하기
    </div>
  </main>
</div>
`;

exports[`SignIn Presentational 전체 화면 체크 2`] = `
<div>
  <main
    class="sc-bdvvtL eNTxIt"
  >
    <h1
      class="sc-gsDKAQ hpyiGY"
    >
      회원가입
    </h1>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        아이디
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="id"
        placeholder="아이디를 10자리 미만으로 기입하세요"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        비밀번호
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="password"
        placeholder="비밀번호를 10자리 미만으로 기입하세요"
        type="password"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        비밀번호 확인
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="passwordCheck"
        placeholder="위와 동일한 비밀번호를 입력하세요"
        type="password"
      />
    </section>
    <section>
      <h5
        class="sc-dkPtRN gaXLmf"
      >
        닉네임
      </h5>
      <input
        class="sc-hKwDye jUfeuU"
        data-testid="nickname"
        placeholder="닉네임을 5자리 미만으로 기입하세요"
      />
    </section>
    <div
      class="sc-eCImPb kWtwVz"
      data-testid="submitButton"
    >
      가입하기
    </div>
  </main>
</div>
`;
